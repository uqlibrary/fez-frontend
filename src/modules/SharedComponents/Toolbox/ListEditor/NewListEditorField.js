import React from 'react';
import PropTypes from 'prop-types';
import NewListEditor from './components/NewListEditor';

const getValue = (input, normalize, searchKey) =>
    normalize(
        (!!input && !!input.value && !!input.value.toJS && input.value.toJS()) ||
            (!!input && !!input.value && input.value) ||
            [],
        searchKey,
    );

export const useItemsList = (input, normalize, searchKey) => {
    const [value, setValue] = React.useState(getValue(input, normalize, searchKey));

    return [value, setValue];
};

export const NewListEditorField = props => {
    const {
        normalize = (value, searchKey) => value.map(item => item[searchKey.value]),
        searchKey = {
            value: 'rek_value',
            order: 'rek_order',
        },
    } = props;

    const [value, setValue] = useItemsList(props.input, normalize, searchKey);

    React.useEffect(() => {
        setValue(getValue(props.input, normalize, searchKey));
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.input]);

    return (
        <NewListEditor
            key={value.length}
            errorText={props.meta ? props.meta.error : null}
            error={props.meta && !!props.meta.error}
            onChange={props.input.onChange}
            remindToAdd={props.remindToAdd}
            list={value}
            searchKey={searchKey}
            normalize={normalize}
            {...props}
        />
    );
};

NewListEditorField.propTypes = {
    searchKey: PropTypes.object,
    normalize: PropTypes.func,
    remindToAdd: PropTypes.bool,
    input: PropTypes.object,
    meta: PropTypes.object,
};

export default React.memo(NewListEditorField);
