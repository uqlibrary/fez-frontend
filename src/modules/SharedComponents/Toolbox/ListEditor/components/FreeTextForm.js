import React, { useCallback, useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import TextField from '@mui/material/TextField';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import makeStyles from '@mui/styles/makeStyles';
import Button from '@mui/material/Button';
import { isValidKeyword } from 'config/validation';
import { indexOf } from 'lodash';

const useStyles = makeStyles(theme => ({
    MUITextLabel: theme.components.MuiFormLabel.styleOverrides,
    remindToAdd: {
        marginTop: 8,
        color: '#f06f0d',
    },
}));

const onItemChangeCallback = (setItem, normalize) => {
    const callback = event => setItem(normalize(event.target.value));

    return [callback, [setItem, normalize]];
};

export const FreeTextForm = ({
    onAdd,
    isValid,
    locale,
    disabled,
    error,
    remindToAdd,
    mode,
    normalize,
    required,
    itemSelectedToEdit,
    listEditorFormId,
    listEditorId,
    onSubmit,
}) => {
    const classes = useStyles();
    const [item, setItem] = useState(itemSelectedToEdit || '');
    const [itemSubmitted, setItemSubmitted] = useState(false);

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const onItemChange = useCallback(...onItemChangeCallback(setItem, normalize));
    const textField = useRef(null);

    useEffect(() => {
        if (itemSubmitted) {
            textField.current.focus();
            setItem('');
            setItemSubmitted(false);
        }

        if (!!itemSelectedToEdit) {
            setItem(itemSelectedToEdit);
        }
    }, [itemSelectedToEdit, itemSubmitted, textField]);

    const addItem = event => {
        // add item if user hits 'enter' key on input field
        if (disabled || (event && event.key && event.key !== 'Enter') || item.length === 0) {
            return;
        }

        // pass on the selected item
        !!onAdd ? onAdd(item) : onSubmit(mode === 'add' ? [item] : item, indexOf);
        setItemSubmitted(true);
    };

    const { inputFieldLabel, inputFieldHint, remindToAddText, addButtonLabel, editButtonLabel } = locale;
    const inputLengthText = isValid(item);
    return (
        <Grid
            container
            spacing={2}
            display="row"
            alignItems="center"
            id={listEditorFormId}
            data-testid={listEditorFormId}
        >
            <Grid item style={{ flexGrow: 1 }}>
                <TextField
                    variant="standard"
                    fullWidth
                    id={`${listEditorId}-input`}
                    data-analyticsid={`${listEditorId}-input`}
                    data-testid={`${listEditorId}-input`}
                    inputProps={{
                        ref: textField,
                    }}
                    label={inputFieldLabel}
                    placeholder={inputFieldHint}
                    value={item}
                    onChange={onItemChange}
                    onKeyDown={addItem}
                    error={!!inputLengthText || (error && item.length === 0)}
                    helperText={inputLengthText}
                    disabled={disabled}
                    required={required}
                />
                {remindToAdd && remindToAddText && item.length !== 0 && !isValid(item) && (
                    <Typography variant="caption" className={classes.remindToAdd}>
                        {remindToAddText}
                    </Typography>
                )}
            </Grid>
            <Grid item xs={12} sm={2}>
                <Button
                    fullWidth
                    id={`${listEditorId}-${mode}`}
                    data-analyticsid={`${listEditorId}-${mode}`}
                    data-testid={`${listEditorId}-${mode}`}
                    color="primary"
                    variant="contained"
                    children={!!itemSelectedToEdit ? editButtonLabel : addButtonLabel}
                    disabled={disabled || item.trim().length === 0 || !!inputLengthText}
                    onClick={addItem}
                />
            </Grid>
        </Grid>
    );
};

FreeTextForm.propTypes = {
    onAdd: PropTypes.func,
    isValid: PropTypes.func,
    locale: PropTypes.object,
    mode: PropTypes.string,
    disabled: PropTypes.bool,
    error: PropTypes.bool,
    remindToAdd: PropTypes.bool,
    normalize: PropTypes.func,
    required: PropTypes.bool,
    itemSelectedToEdit: PropTypes.any,
    listEditorFormId: PropTypes.string,
    listEditorId: PropTypes.string,
    onSubmit: PropTypes.func,
};

FreeTextForm.defaultProps = {
    isValid: isValidKeyword(2000),
    remindToAdd: false,
    locale: {
        id: 'free-text-input',
        inputFieldLabel: 'Item name',
        inputFieldHint: 'Please type the item name',
        addButtonLabel: 'Add',
    },
    required: false,
    itemSelectedToEdit: '',
    mode: 'add',
};

export default React.memo(FreeTextForm);
