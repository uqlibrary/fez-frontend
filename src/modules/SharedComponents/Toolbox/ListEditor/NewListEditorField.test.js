import React from 'react';
import NewListEditorField from './NewListEditorField';
import { List } from 'immutable';
import { rtlRender } from 'test-utils';
import { GenericTemplate } from './components/GenericTemplate';
import { FreeTextForm } from './components/FreeTextForm';

describe('NewListEditorField component', () => {
    it('should render new list editor component with the given array', () => {
        const props = {
            meta: {
                error: 'test1',
            },
            input: {
                onChange: jest.fn(),
                value: [
                    { rek_keywords: 'test', rek_keywords_order: 1 },
                    { rek_keywords: 'testing', rek_keywords_order: 2 },
                ],
            },
            remindToAdd: true,
            maxInputLength: 100,
            searchKey: {
                value: 'rek_keywords',
                order: 'rek_keywords_order',
            },
            listEditorId: 'test',
            ListEditorForm: FreeTextForm,
            ListEditorItemTemplate: GenericTemplate,
        };
        const { asFragment } = rtlRender(<NewListEditorField {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render new list editor component with the given List', () => {
        const props = {
            meta: {
                error: 'test1',
            },
            input: {
                onChange: jest.fn(),
                value: new List([
                    { rek_keywords: 'test', rek_keywords_order: 1 },
                    { rek_keywords: 'testing', rek_keywords_order: 2 },
                ]),
            },
            remindToAdd: true,
            maxInputLength: 100,
            searchKey: {
                value: 'rek_keywords',
                order: 'rek_keywords_order',
            },
            listEditorId: 'test',
            ListEditorForm: FreeTextForm,
            ListEditorItemTemplate: GenericTemplate,
        };
        const { asFragment } = rtlRender(<NewListEditorField {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render new list editor component without any list items', () => {
        const props = {
            input: {
                onChange: jest.fn(),
            },
            remindToAdd: true,
            maxInputLength: 100,
            searchKey: {
                value: 'rek_keywords',
                order: 'rek_keywords_order',
            },
            listEditorId: 'test',
            ListEditorForm: FreeTextForm,
            ListEditorItemTemplate: GenericTemplate,
        };
        const { asFragment } = rtlRender(<NewListEditorField {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
