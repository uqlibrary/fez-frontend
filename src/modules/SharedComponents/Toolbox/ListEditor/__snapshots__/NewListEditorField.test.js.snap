// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewListEditorField component should render new list editor component with the given List 1`] = `
<DocumentFragment>
  <div
    id="test-list-editor"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-hckusd-MuiGrid-root"
      data-testid="test-form-add"
      id="test-form-add"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        style="flex-grow: 1;"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-h9aup-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="test-input"
            id="test-input-label"
          >
            Item name
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1vg8b0p-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              data-analyticsid="test-input"
              data-testid="test-input"
              id="test-input"
              placeholder="Please type the item name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 css-63b4os-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1duotrm-MuiButtonBase-root-MuiButton-root"
          data-analyticsid="test-add"
          data-testid="test-add"
          disabled=""
          id="test-add"
          tabindex="-1"
          type="button"
        >
          Add
        </button>
      </div>
    </div>
    <div
      style="flex-grow: 1; padding: 8px;"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-19d556j-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-6 css-fyaqcr-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1e4ju4k-MuiTypography-root"
          >
            Name
          </span>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-5 css-1olnuyg-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1e4ju4k-MuiTypography-root"
          >
            Reorder items
          </span>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-sm-1 css-1lgqump-MuiGrid-root"
        >
          <span
            aria-label="Remove all items"
            class=""
            data-mui-internal-clone-element="true"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
              data-analyticsid="delete-all-test"
              data-testid="delete-all-test"
              id="delete-all-test"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteForeverIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </button>
          </span>
        </div>
      </div>
    </div>
    <div
      data-testid="test-list"
      id="test-list"
    >
      <div
        data-testid="test-list-row-0"
        id="test-list-row-0"
        style="flex-grow: 1; padding: 8px;"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1x0yq54-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-6 MuiGrid-grid-md-9 css-1ncsxfs-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-1omq3q2-MuiTypography-root"
            >
              test
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-5 MuiGrid-grid-md-2 css-o9thq8-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <span
                  aria-label="Move item down the order"
                  class=""
                  data-mui-internal-clone-element="true"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
                    data-analyticsid="test-list-row-0-move-down"
                    data-testid="test-list-row-0-move-down"
                    id="test-list-row-0-move-down"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="KeyboardArrowDownIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-sm-1 css-1lgqump-MuiGrid-root"
          >
            <span
              aria-label="Remove this item"
              class=""
              data-mui-internal-clone-element="true"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
                data-analyticsid="test-list-row-0-delete"
                data-testid="test-list-row-0-delete"
                id="test-list-row-0-delete"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        data-testid="test-list-row-1"
        id="test-list-row-1"
        style="flex-grow: 1; padding: 8px;"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1x0yq54-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-6 MuiGrid-grid-md-9 css-1ncsxfs-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-1omq3q2-MuiTypography-root"
            >
              testing
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-5 MuiGrid-grid-md-2 css-o9thq8-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <span
                  aria-label="Move item up the order"
                  class=""
                  data-mui-internal-clone-element="true"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
                    data-analyticsid="test-list-row-1-move-up"
                    data-testid="test-list-row-1-move-up"
                    id="test-list-row-1-move-up"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="KeyboardArrowUpIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7.41 15.41 12 10.83l4.59 4.58L18 14l-6-6-6 6z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-sm-1 css-1lgqump-MuiGrid-root"
          >
            <span
              aria-label="Remove this item"
              class=""
              data-mui-internal-clone-element="true"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
                data-analyticsid="test-list-row-1-delete"
                data-testid="test-list-row-1-delete"
                id="test-list-row-1-delete"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
    <p
      class="MuiFormHelperText-root Mui-error css-1d7hrfi-MuiFormHelperText-root"
    >
      test1
    </p>
  </div>
</DocumentFragment>
`;

exports[`NewListEditorField component should render new list editor component with the given array 1`] = `
<DocumentFragment>
  <div
    id="test-list-editor"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-hckusd-MuiGrid-root"
      data-testid="test-form-add"
      id="test-form-add"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        style="flex-grow: 1;"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-h9aup-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="test-input"
            id="test-input-label"
          >
            Item name
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1vg8b0p-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              data-analyticsid="test-input"
              data-testid="test-input"
              id="test-input"
              placeholder="Please type the item name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 css-63b4os-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1duotrm-MuiButtonBase-root-MuiButton-root"
          data-analyticsid="test-add"
          data-testid="test-add"
          disabled=""
          id="test-add"
          tabindex="-1"
          type="button"
        >
          Add
        </button>
      </div>
    </div>
    <div
      style="flex-grow: 1; padding: 8px;"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-19d556j-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-6 css-fyaqcr-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1e4ju4k-MuiTypography-root"
          >
            Name
          </span>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-5 css-1olnuyg-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1e4ju4k-MuiTypography-root"
          >
            Reorder items
          </span>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-sm-1 css-1lgqump-MuiGrid-root"
        >
          <span
            aria-label="Remove all items"
            class=""
            data-mui-internal-clone-element="true"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
              data-analyticsid="delete-all-test"
              data-testid="delete-all-test"
              id="delete-all-test"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteForeverIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </button>
          </span>
        </div>
      </div>
    </div>
    <div
      data-testid="test-list"
      id="test-list"
    >
      <div
        data-testid="test-list-row-0"
        id="test-list-row-0"
        style="flex-grow: 1; padding: 8px;"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1x0yq54-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-6 MuiGrid-grid-md-9 css-1ncsxfs-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-1omq3q2-MuiTypography-root"
            >
              test
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-5 MuiGrid-grid-md-2 css-o9thq8-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <span
                  aria-label="Move item down the order"
                  class=""
                  data-mui-internal-clone-element="true"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
                    data-analyticsid="test-list-row-0-move-down"
                    data-testid="test-list-row-0-move-down"
                    id="test-list-row-0-move-down"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="KeyboardArrowDownIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-sm-1 css-1lgqump-MuiGrid-root"
          >
            <span
              aria-label="Remove this item"
              class=""
              data-mui-internal-clone-element="true"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
                data-analyticsid="test-list-row-0-delete"
                data-testid="test-list-row-0-delete"
                id="test-list-row-0-delete"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        data-testid="test-list-row-1"
        id="test-list-row-1"
        style="flex-grow: 1; padding: 8px;"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1x0yq54-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-6 MuiGrid-grid-md-9 css-1ncsxfs-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-1omq3q2-MuiTypography-root"
            >
              testing
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-5 MuiGrid-grid-md-2 css-o9thq8-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <span
                  aria-label="Move item up the order"
                  class=""
                  data-mui-internal-clone-element="true"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
                    data-analyticsid="test-list-row-1-move-up"
                    data-testid="test-list-row-1-move-up"
                    id="test-list-row-1-move-up"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="KeyboardArrowUpIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7.41 15.41 12 10.83l4.59 4.58L18 14l-6-6-6 6z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-sm-1 css-1lgqump-MuiGrid-root"
          >
            <span
              aria-label="Remove this item"
              class=""
              data-mui-internal-clone-element="true"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-bc0omg-MuiButtonBase-root-MuiIconButton-root"
                data-analyticsid="test-list-row-1-delete"
                data-testid="test-list-row-1-delete"
                id="test-list-row-1-delete"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
    <p
      class="MuiFormHelperText-root Mui-error css-1d7hrfi-MuiFormHelperText-root"
    >
      test1
    </p>
  </div>
</DocumentFragment>
`;

exports[`NewListEditorField component should render new list editor component without any list items 1`] = `
<DocumentFragment>
  <div
    id="test-list-editor"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-hckusd-MuiGrid-root"
      data-testid="test-form-add"
      id="test-form-add"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        style="flex-grow: 1;"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-h9aup-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="test-input"
            id="test-input-label"
          >
            Item name
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1vg8b0p-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              data-analyticsid="test-input"
              data-testid="test-input"
              id="test-input"
              placeholder="Please type the item name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 css-63b4os-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1duotrm-MuiButtonBase-root-MuiButton-root"
          data-analyticsid="test-add"
          data-testid="test-add"
          disabled=""
          id="test-add"
          tabindex="-1"
          type="button"
        >
          Add
        </button>
      </div>
    </div>
    <div
      data-testid="test-list"
      id="test-list"
    />
    <p
      class="MuiFormHelperText-root css-1d7hrfi-MuiFormHelperText-root"
    />
  </div>
</DocumentFragment>
`;
